# Stock Market Research Platform

## Exploratory Data Analysis of Technology Stocks

### Author
Rashika Agarwal

### Objective

The objective of this notebook is to perform exploratory data analysis on historical stock market data collected using Yahoo Finance.

This analysis aims to:

- Understand stock price behavior over time
- Explore the statistical properties of the data
- Identify missing values and potential data quality issues
- Generate insights using visualizations
- Prepare the dataset for feature engineering and machine learning

# Exploratory Data Analysis

## 1. Import Required Libraries

In [10]:
import pandas as pd
import numpy as np

import plotly.express as px
import plotly.graph_objects as go

## 2. Load Apple Stock Dataset

In [9]:
aapl = pd.read_csv(
    "../data/raw/AAPL.csv",
    index_col="Date",
    parse_dates=True
)

aapl.head()

,Close,High,Low,Open,Volume
Date,,,,,
2020-01-02,72.333885,72.394093,71.091191,71.344062,135480400
2020-01-03,71.630661,72.389280,71.406689,71.563228,146322800
2020-01-06,72.201431,72.239965,70.503569,70.754036,118387200
2020-01-07,71.861839,72.466322,71.642681,72.211041,108872000
2020-01-08,73.017853,73.318893,71.565636,71.565636,132079200


## 3. Dataset Information

In [11]:
aapl.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1628 entries, 2020-01-02 to 2026-06-25
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Close   1628 non-null   float64
 1   High    1628 non-null   float64
 2   Low     1628 non-null   float64
 3   Open    1628 non-null   float64
 4   Volume  1628 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 76.3 KB


## 4. Statistical Summary

In [12]:
aapl.describe()

,Close,High,Low,Open,Volume
count,1628.000000,1628.000000,1628.000000,1628.000000,1.628000e+03
mean,172.603195,174.390870,170.683961,172.462315,8.188450e+07
std,55.737273,56.179436,55.327491,55.738475,5.049625e+07
min,54.163704,55.160702,51.324800,55.059313,1.791060e+07
25%,134.593174,136.302051,132.415615,134.507080,4.890652e+07
50%,168.251762,169.689272,166.719160,167.915708,6.793505e+07
75%,213.077183,215.308501,211.005907,213.064585,9.690462e+07
max,315.200012,317.399994,309.649994,314.179993,4.265100e+08


## 5. Missing Value Analysis

In [13]:
aapl.isnull().sum()

Close     0
High      0
Low       0
Open      0
Volume    0
dtype: int64

## 6. Dataset Dimensions

In [14]:
aapl.shape

(1628, 5)

## 7. Column Names

In [15]:
aapl.columns

Index(['Close', 'High', 'Low', 'Open', 'Volume'], dtype='object')

## 8. Date Index

In [16]:
aapl.index

DatetimeIndex(['2020-01-02', '2020-01-03', '2020-01-06', '2020-01-07',
               '2020-01-08', '2020-01-09', '2020-01-10', '2020-01-13',
               '2020-01-14', '2020-01-15',
               ...
               '2026-06-11', '2026-06-12', '2026-06-15', '2026-06-16',
               '2026-06-17', '2026-06-18', '2026-06-22', '2026-06-23',
               '2026-06-24', '2026-06-25'],
              dtype='datetime64[ns]', name='Date', length=1628, freq=None)

# Visual Exploratory Data Analysis

## 9. Apple Closing Price Trend
The closing price represents the final trading price of Apple stock for each trading day.

Visualizing the closing price over time helps identify long-term growth, major market movements, and periods of increased volatility.

In [17]:
fig = px.line(
    aapl,
    x=aapl.index,
    y="Close",
    title="Apple Closing Price Over Time"
)

fig.update_layout(
    xaxis_title="Date",
    yaxis_title="Closing Price (USD)",
    template="plotly_white"
)

fig.show()

### Key Insights
- Apple's closing price exhibits a clear long-term upward trend from 2020 to 2026, indicating sustained growth over the analysis period.
- Although several short-term price corrections are visible, the stock consistently recovered and continued its upward trajectory.
- Price fluctuations become more noticeable during 2025–2026, suggesting periods of increased market volatility.
- The strong long-term trend indicates that Apple's historical price data is well suited for time-series forecasting and trend-based predictive modeling.

Overall, Apple demonstrates a strong long-term upward trend with periodic corrections, making historical prices suitable for subsequent forecasting analysis.

---
## 10. Apple Trading Volume

Trading volume represents the number of Apple shares traded each day.

Higher trading volume often indicates increased investor activity and may coincide with important financial events or earnings announcements.

In [25]:
fig = px.line(
    aapl,
    x=aapl.index,
    y="Volume",
    title="Apple Daily Trading Volume"
)

fig.update_layout(
    xaxis_title="Date",
    yaxis_title="Volume",
    template="plotly_white"
)

fig.show()

### Key Insights
- Trading volume was highest during the early part of the dataset, particularly in 2020, indicating elevated investor activity.
- After 2020, trading volume gradually stabilized while continuing to exhibit occasional sharp spikes.
- Large spikes in trading volume likely correspond to earnings announcements, product launches, or major market events.
- The combination of relatively stable trading activity and occasional high-volume spikes suggests consistent investor participation punctuated by periods of heightened market attention.

Overall, trading activity remained stable over time while reacting sharply to significant market events.

---
## 11. Apple Closing Price Distribution

A histogram illustrates how frequently different closing prices occurred during the selected time period.

This helps identify common price ranges and understand the overall distribution of closing prices.

In [26]:
fig = px.histogram(
    aapl,
    x="Close",
    nbins=30,
    title="Distribution of Apple Closing Prices"
)

fig.update_layout(
    xaxis_title="Closing Price (USD)",
    yaxis_title="Frequency",
    template="plotly_white"
)

fig.show()

### Key Insights
- Most closing prices are concentrated between approximately **120 USD and 190 USD**, representing the price range where Apple traded most frequently. 
- The distribution extends toward higher prices above **250 USD**, reflecting Apple's strong long-term appreciation during the analysis period.
- The histogram exhibits a **moderate right skew**, indicating that higher closing prices occurred less frequently than lower price levels.
- The absence of multiple distinct peaks indicates that Apple's price evolved gradually throughout the analysis period rather than experiencing abrupt structural shifts.

Overall, the distribution reflects Apple's gradual long-term appreciation in stock price.

---
## 12. Apple Closing Price Box Plot

A box plot summarizes the distribution of Apple's closing prices by displaying the median, quartiles, and potential outliers.

In [20]:
fig = px.box(
    aapl,
    y="Close",
    title="Apple Closing Price Box Plot"
)

fig.update_layout(
    yaxis_title="Closing Price (USD)",
    template="plotly_white"
)

fig.show()

### Key Insights
- The median closing price is approximately **168.25 USD**, indicating the central value of Apple's stock price during the analysis period.
- The middle 50% of closing prices (interquartile range) lie between **134.54 USD (Q1)** and **213.12 USD (Q3)**, showing that most trading prices were concentrated within this range.
- The stock price varied substantially over time, ranging from approximately **54.16 USD** to **315.20 USD**, reflecting Apple's long-term growth.
- No significant statistical outliers were identified using the IQR rule, suggesting that the distribution is relatively stable despite substantial long-term price growth.

Overall, the box plot confirms a relatively stable distribution without extreme outliers.

---
## 13. Apple Daily Percentage Returns

Daily return measures the percentage change in Apple's closing price from one trading day to the next.

It provides a standardized way to evaluate stock performance and market volatility.

In [21]:
aapl["Daily Return"] = aapl["Close"].pct_change()

aapl.head()

,Close,High,Low,Open,Volume,Daily Return
Date,,,,,,
2020-01-02,72.333885,72.394093,71.091191,71.344062,135480400,NaN
2020-01-03,71.630661,72.389280,71.406689,71.563228,146322800,-0.009722
2020-01-06,72.201431,72.239965,70.503569,70.754036,118387200,0.007968
2020-01-07,71.861839,72.466322,71.642681,72.211041,108872000,-0.004703
2020-01-08,73.017853,73.318893,71.565636,71.565636,132079200,0.016087


### Explanation

- Daily percentage returns were calculated using the percentage change between consecutive closing prices.
- This derived feature will be used throughout the remaining exploratory analysis to examine market volatility and return distributions.

---
## 14. Apple Daily Return Trend

This visualization displays Apple's daily percentage returns over time.

Large positive or negative spikes indicate periods of increased market volatility.

In [22]:
fig = px.line(
    aapl,
    x=aapl.index,
    y="Daily Return",
    title="Apple Daily Percentage Returns"
)

fig.update_layout(
    xaxis_title="Date",
    yaxis_title="Daily Return",
    template="plotly_white"
)

fig.show()

### Key Insights
- Daily returns fluctuate around **zero**, indicating that most day-to-day price changes are relatively small.
- Short periods of unusually large positive and negative returns indicate episodes of increased market volatility.
- Extreme daily returns occur infrequently, suggesting that substantial gains or losses are relatively rare.
- Since daily returns do not exhibit a persistent trend, predictive models should focus on short-term market dynamics rather than assuming long-term directional movement.

---
## 15. Apple Daily Return Distribution

The histogram below illustrates how Apple's daily percentage returns are distributed.

Most observations are expected to cluster around zero, with fewer extreme positive or negative returns.

In [24]:
fig = px.histogram(
    aapl,
    x="Daily Return",
    nbins=60,
    title="Distribution of Apple Daily Returns"
)

fig.update_layout(
    xaxis_title="Daily Return",
    yaxis_title="Frequency",
    template="plotly_white"
)

fig.show()

### Key Insights

- The distribution is centered close to **0%**, indicating that most trading days experience relatively small price changes.
- The histogram resembles an approximately bell-shaped distribution, with the highest concentration of returns occurring near zero.
- Large positive and negative daily returns appear infrequently and are confined to the tails of the distribution.
- The presence of both positive and negative extreme returns indicates occasional periods of elevated market volatility, while the concentration around zero confirms that most daily price movements are relatively small.

Overall, daily returns fluctuate around zero with occasional volatility spikes, which is typical of financial time-series data.